### PR TITLE
Add check for consistency between nActionsOrchard and Orchard flags.

### LINKED ISSUE #5245: Implement consensus rule that if effectiveVersion ≥ 5 and nActionsOrchard > 0, at least one of enableSpendsOrchard and enableOutputsOrchard is 1
https://github.com/zcash/zips/issues/531

### PR DESCRIPTION
**Pull request by nuttycom**
Fixes #5245

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -1134,6 +1133,13 @@ bool ContextualCheckTransaction(
                 REJECT_INVALID, "bad-tx-too-many-orchard-actions");
         }
 
+        if (orchard_bundle.GetNumActions() > 0 && !orchard_bundle.OutputsEnabled() && !orchard_bundle.SpendsEnabled()) {
+            return state.DoS(
+                dosLevelPotentiallyRelaxing,
+                error("ContextualCheckTransaction(): Orchard actions are present, but flags do not permit Orchard spends or outputs"),
+                REJECT_INVALID, "bad-tx-orchard-flags-disable-actions");
+        }
```
Comment (daira) [inline src/main.cpp:1141]: This should not be a contextual check.

**Review by daira — CHANGES_REQUESTED** [review]
Move this to be a non-contextual check.

File: src/main.cpp (hunk 1 above)
Comment (daira) [reply src/main.cpp:1141]: The restrictions on `vShieldedSpend.size()`, `vShieldedOutput.size()`, and `GetNumActions()` also don't need to be contextual (the former two are weaker than what is already implied by the 2 MB block size limit).

Comment (daira) [reply src/main.cpp:1141]: Also, the check at https://github.com/zcash/zcash/blob/5802be342781b7b7d37f61b70f333fbb14ab5409/src/main.cpp#L1510-L1514 is made redundant by this check (and therefore is no longer in the spec).

**Comment by CRYONE7** [comment]
MTZrb1ZVcFFyVnI4Wm90VXI5ZzNSUTdBZ1dOdXdwWVZ3cw==

File: src/main.cpp [hunk 2]
```diff
@@ -1505,14 +1483,39 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
         }
     }
 
-    auto valueBalanceOrchard = orchard_bundle.GetValueBalance();
+    // nSpendsSapling, nOutputsSapling, and nActionsOrchard MUST all be less than 2^16
+    size_t max_elements = (1 << 16) - 1;
+    if (tx.vShieldedSpend.size() > max_elements) {
+        return state.DoS(
+            dosLevelPotentiallyRelaxing,
```
Comment (str4d) [inline src/main.cpp:1490]: `dosLevelPotentiallyRelaxing` is only relevant to contextual rules (that can change with block height), so this isn't defined here (and is causing the build to fail):
```suggestion
            100,
```

File: src/main.cpp [hunk 3]
```diff
@@ -1505,14 +1483,39 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
...
+            error("ContextualCheckTransaction(): 2^16 or more Sapling spends"),
+            REJECT_INVALID, "bad-tx-too-many-sapling-spends");
+    }
+    if (tx.vShieldedOutput.size() > max_elements) {
+        return state.DoS(
+            dosLevelPotentiallyRelaxing,
```
Comment (str4d) [inline src/main.cpp:1496]:
```suggestion
            100,
```

File: src/main.cpp [hunk 4]
```diff
@@ -1505,14 +1483,39 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
...
+            error("ContextualCheckTransaction(): 2^16 or more Sapling outputs"),
+            REJECT_INVALID, "bad-tx-too-many-sapling-outputs");
+    }
+    if (orchard_bundle.GetNumActions() > max_elements) {
+        return state.DoS(
+            dosLevelPotentiallyRelaxing,
```
Comment (str4d) [inline src/main.cpp:1502]:
```suggestion
            100,
```

**Review by str4d — CHANGES_REQUESTED** [review]

File: src/main.cpp [hunk 5]
```diff
@@ -1505,14 +1483,39 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
         }
     }
 
-    auto valueBalanceOrchard = orchard_bundle.GetValueBalance();
+    // nSpendsSapling, nOutputsSapling, and nActionsOrchard MUST all be less than 2^16
+    size_t max_elements = (1 << 16) - 1;
+    if (tx.vShieldedSpend.size() > max_elements) {
+        return state.DoS(
+            dosLevelPotentiallyRelaxing,
+            error("ContextualCheckTransaction(): 2^16 or more Sapling spends"),
```
Comment (daira) [inline src/main.cpp:1491]:
```suggestion
            error("CheckTransactionWithoutProofVerification(): 2^16 or more Sapling spends"),
```

File: src/main.cpp [hunk 6]
```diff
@@ -1505,14 +1483,39 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
...
+            REJECT_INVALID, "bad-tx-too-many-sapling-spends");
+    }
+    if (tx.vShieldedOutput.size() > max_elements) {
+        return state.DoS(
+            dosLevelPotentiallyRelaxing,
+            error("ContextualCheckTransaction(): 2^16 or more Sapling outputs"),
```
Comment (daira) [inline src/main.cpp:1497]:
```suggestion
            error("CheckTransactionWithoutProofVerification(): 2^16 or more Sapling outputs"),
```

File: src/main.cpp [hunk 7]
```diff
@@ -1505,14 +1483,39 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
...
+            REJECT_INVALID, "bad-tx-too-many-sapling-outputs");
+    }
+    if (orchard_bundle.GetNumActions() > max_elements) {
+        return state.DoS(
+            dosLevelPotentiallyRelaxing,
+            error("ContextualCheckTransaction(): 2^16 or more Orchard actions"),
```
Comment (daira) [inline src/main.cpp:1503]:
```suggestion
            error("CheckTransactionWithoutProofVerification(): 2^16 or more Orchard actions"),
```

File: src/main.cpp [hunk 8]
```diff
@@ -1505,14 +1483,39 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
...
-    if (!orchard_bundle.SpendsEnabled() && !orchard_bundle.OutputsEnabled() && valueBalanceOrchard != 0) {
-        return state.DoS(100, error("CheckTransaction(): tx.valueBalanceOrchard has no sources or sinks"),
-                         REJECT_INVALID, "bad-txns-valuebalance-nonzero");
+    // Check that if neither Orchard spends nor outputs are enabled, the transaction contains
+    // no Orchard actions. This subsumes the check that valueBalanceOrchard must equal zero
+    // in the case that both spends and outputs are disabled.
+    if (orchard_bundle.GetNumActions() > 0 && !orchard_bundle.OutputsEnabled() && !orchard_bundle.SpendsEnabled()) {
+        return state.DoS(
+            100,
+            error("ContextualCheckTransaction(): Orchard actions are present, but flags do not permit Orchard spends or outputs"),
```
Comment (daira) [inline src/main.cpp:1513]:
```suggestion
            error("CheckTransactionWithoutProofVerification(): Orchard actions are present, but flags do not permit Orchard spends or outputs"),
```

**Review by daira — APPROVED** [review]
utACK modulo my and @str4d's comments.

File: src/main.cpp (hunk 5 above)
Comment (str4d) [reply src/main.cpp:1491]: I'd rather we change this to `CheckTransaction`. The only reason `*WithoutProofVerification` exists is a Sprout artifact that I want to refactor away, after which we could undo the split.

**Review by daira — APPROVED** [review]
utACK 660686d412b67f4206219d4d29cb234dedc68865

**Review by str4d — APPROVED** [review]
utACK

**Comment by str4d** [comment]
@zkbot r+

**Comment by str4d** [comment]
Transient CI failure.

@zkbot retry

**Comment by str4d** [comment]
@zkbot retry

**Comment by str4d** [comment]
Ugh

@zkbot retry

**Pull request merged**